src/persist/src/indexed/mod.rs

```diff
@@ -484,6 +485,31 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         ret
     }
 
+    /// Compact all traces, if they have batches that can be compacted.
+    ///
+    /// TODO: currently we do not attempt to compact future batches and instead
+    /// logically delete them from future after all updates contained within a
+    /// given future batch have been moved over to trace. This policy works fine
+    /// assuming data mostly arrives in order, or not very far in advance of the
+    /// currently sealed time. We will need to revisit the future compaction if
+    /// that assumption stops being true.
+    fn compact(&mut self) -> Result<(), Error> {
```

Comment:
oh this isn't quite what I meant. I was saying that we should only have the restore handling in one place instead of once per trace.step call. something like for this method:
``` 
       for (_, trace) in self.traces.iter_mut() {
            trace.step(&mut self.blob)?;
        }
```
and then the restore/try_set_meta part in the caller

Reply:
oh shoot i just saw this. I misunderstood the original comment. I'm happy to have a `compact_inner` method do the `trace.steps` and then have `compact` call `compact_inner` and based on the return value either restore or try to commit metadata.

I like having the various things in `step` (currently `drain_pending` and `compact`) both appropriately clean up after themselves as it makes the top level of step easier to reason about.

Let me know if that sounds worth doing or if you'd rather I take a different approach?

Reply:
That seems fine (and better than my suggestion, dunno why I hesitate to add 2 functions, it's not like we have a budget for them)

Reply:
It's also fine to just ignore this, the stuff more directly related to getting stability testing passing is much more important. If I'd really cared, I'd have disabled your auto-merge :D